src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1345,9 +1411,16 @@ public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
                     .WithParseOptions(existingGeneratedState.ParseOptions)
                     .WithGenerationDateTime(generationDateTime);
 
+                if (syntaxNode != null)
+                {
+                    newGeneratedState = newGeneratedState.WithSyntaxRoot(syntaxNode);
+                }
```

Comment:
i dont' quite get this.  why are we only setting the node if not null.  if it was null, shouldn't we set the newGeneratedState to have a null node?  Could this lead to a sitaution where we hold onto old node state inappropriately?  (if this reveals a bug, please add a test that would help catch this).

Reply:
Calling `WithText` on `newGeneratedState` means it will have a lazily evaluated tree, it doesn't reuse the old tree. Could move the null check into `WithSyntaxRoot` and make it a no-op though, to make the code nicer to look at?

Reply:
I'm still scratching my head here -- I'd expect the source text and tree are both nullable, but exactly one is non-null. Then there's a single call to either WithText() or WithSyntaxRoot() that matches which was nullable.